bundles/org.dataflowanalysis.analysis.converter/src/org/dataflowanalysis/analysis/converter/PalladioConverter.java

```diff
@@ -0,0 +1,184 @@
+
+            if (optFlow.isPresent()) {
+                return;
+            }
+
+            Flow newFlow = dataflowdiagramFactory.eINSTANCE.createFlow();
+            newFlow.setSourceNode(source);
+            newFlow.setDestinationNode(dest);
+            newFlow.setEntityName(flowName);
+
+            // Palladio Assumption: Each flows between two nodes with the same parameters/that are called the same use the same pin
+            Pin sourceOutPin = findOrCreateOutputPin(source, flowName);
+            Pin destInPin = findOrCreateInputPin(dest, flowName);
+            newFlow.setSourcePin(sourceOutPin);
+            newFlow.setDestinationPin(destInPin);
+
+            ForwardingAssignment forwarding = datadictionaryFactory.eINSTANCE.createForwardingAssignment();
+            forwarding.setOutputPin(sourceOutPin);
+            source.getBehaviour().getAssignment().add(forwarding);
+
+            this.dataFlowDiagram.getFlows().add(newFlow);
+        }
+    }
+
+    // A pin is equivalent if the same parameters are passed
+    private Pin findOrCreateOutputPin(Node source, String parameters) {
+        Optional<Pin> optPin = source.getBehaviour().getOutPin().stream().filter(p -> p.getEntityName().equals(parameters)).findAny();
+        if (optPin.isPresent()) {
+            return optPin.get();
+        }
+        Pin pin = datadictionaryFactory.eINSTANCE.createPin();
+        pin.setEntityName(parameters);
+        source.getBehaviour().getOutPin().add(pin);
+        return pin;
+    }
+
+    // A pin is equivalent if the same parameters are passed
+    private Pin findOrCreateInputPin(Node dest, String parameters) {
+        Optional<Pin> optPin = dest.getBehaviour().getInPin().stream().filter(p -> p.getEntityName().equals(parameters)).findAny();
+        if (optPin.isPresent()) {
+            return optPin.get();
+        }
+        Pin pin = datadictionaryFactory.eINSTANCE.createPin();
+        pin.setEntityName(parameters);
+        dest.getBehaviour().getInPin().add(pin);
+        return pin;
+    }
+
+    private Node getOrCreateDFDNode(AbstractPCMActionSequenceElement<? extends Entity> pcmASE) {
+        Node dfdNode = dfdNodeMap.get(pcmASE.getElement());
+
+        if (dfdNode == null) {
+            dfdNode = createCorrespondingDFDNode(pcmASE);
+            dfdNodeMap.put(pcmASE.getElement(), dfdNode);
+        }
+
+        addNodeCharacteristicsToNode(dfdNode, pcmASE.getAllNodeCharacteristics());
+
+        return dfdNode;
+    }
+
+    private Node createCorrespondingDFDNode(AbstractPCMActionSequenceElement<? extends Entity> pcmASE) {
+        Node node;
+
+        if (pcmASE instanceof UserActionSequenceElement<?>) {
+            node = dataflowdiagramFactory.eINSTANCE.createExternal();
+        } else if (pcmASE instanceof SEFFActionSequenceElement<?>) {
+            node = dataflowdiagramFactory.eINSTANCE.createProcess();
+        } else {
+            return null;
```

Comment:
Log an error here

Reply:
What would the better behavior be here? Throw an Exception to indicate a parsing error, or returning an Optional and skipping the node?